diff --git a/godot-mesa/src/microsoft/compiler/nir_to_dxil.c b/godot-mesa/src/microsoft/compiler/nir_to_dxil.c
index d6c9d06..1e8e9ea 100644
--- a/godot-mesa/src/microsoft/compiler/nir_to_dxil.c
+++ b/godot-mesa/src/microsoft/compiler/nir_to_dxil.c
@@ -5083,11 +5083,11 @@ emit_intrinsic(struct ntd_context *ctx, nir_intrinsic_instr *intr)
    case nir_intrinsic_exclusive_scan:
       return emit_reduce(ctx, intr);
 
-   case nir_intrinsic_load_constant_non_opt:
+   case nir_intrinsic_load_constant_non_opt: {
       const struct dxil_value* value = get_src(ctx, &intr->src[0], 0, nir_type_uint);
       store_dest_value(ctx, &intr->dest, 0, value);
       return true;
-
+   }
    case nir_intrinsic_load_num_workgroups:
    case nir_intrinsic_load_workgroup_size:
    default:
